File: src/clawer.py
import asyncio
import itertools
import json
import os
import re
from collections import Counter
from datetime import date

import requests
from bs4 import BeautifulSoup


BASE_URL = 'https://egov.uscis.gov/casestatus/mycasestatus.do?appReceiptNum='

# see https://egov.uscis.gov/processing-times/
FORM_TYPES = [
    'I-90', 'I-102', 'I-129', 'I-129CW', 'I-129F', 'I-130', 'I-131', 'I-140', 'I-212', 'I-360', 'I-485', 'I-526',
    'I-539', 'I-600', 'I-600A', 'I-601', 'I-601A', 'I-612', 'I-730', 'I-751', 'I-765', 'I-765V', 'I-800', 'I-800A',
    'I-817', 'I-821', 'I-821D', 'I-824', 'I-829', 'I-914', 'I-918', 'I-924', 'I-929'
]
CENTER_NAMES = ['WAC', 'EAC', 'VSC',  'CSC', 'LIN',
                'NSC', 'SRC', 'TSC', 'MSC', 'NBC', 'IOE', 'YSC']


def get_case_id(center_name: str, two_digit_yr: int, day: int, code: int, case_serial_numbers: int) -> str:
    return center_name + str(two_digit_yr) + str(day) + str(code) + (str(case_serial_numbers)).zfill(4)


def get_form_type(text: str):
    for form in FORM_TYPES:
        if form in text:
            return form
    return 'unknown form type'


def parse_response(response):
    text = str(BeautifulSoup(response.text, 'html.parser').findAll(
        'div', {'class': 'text-center'})[0])
    status = re.search('(?<=<h1>).*(?=</h1>)', text)
    if status is None or status[0] == '' or status.group(0) == '':
        return None
    else:
        return status.group(0), get_form_type(text)


def get_last_case_number(center_name: str, two_digit_yr: int, day: int, code: int):
    if parse_response(requests.get(BASE_URL + get_case_id(center_name, two_digit_yr, day, code, 1))) is None:
        return 0
    low = 1
    high = 9999
    while low < high:
        mid = int((low + high) / 2)
        result = parse_response(requests.get(
            BASE_URL + get_case_id(center_name, two_digit_yr, day, code, mid)))
        # find first empty response
        if result is not None:
            low = mid + 1
        else:
            high = mid
    return low - 1


def merge(counter: Counter):
    current_day_since_1970 = str((date.today() - date(1970, 1, 1)).days)
    file_path = os.path.dirname(os.path.realpath(__file__)) + '/data.json'
    with open(file_path) as f:
        counter_all_days = json.loads(f.read())
    with open(file_path, 'w') as f:
        for key in counter:
            if key not in counter_all_days:
                counter_all_days[key] = {}
            counter_all_days[key][current_day_since_1970] = counter[key]
        f.write(json.dumps(counter_all_days, sort_keys=True, indent=4))


def request_ignore_err(url: str):
    try:
        return requests.get(url)
    except:
        print(f'failed in {url}')
        return None


async def claw(center_name: str, two_digit_yr: int, day: int, code: int, counter: Counter):
    event_loop = asyncio.get_event_loop()

    case_serial_numbers = range(1, get_last_case_number(
        center_name, two_digit_yr, day, code))
    ids = [get_case_id(center_name, two_digit_yr, day, code, number)
           for number in case_serial_numbers]
    print(
        f'clawing center {center_name}, day: {day}, code: {code}, total of {len(ids)} cases')
    futures = [event_loop.run_in_executor(
        None, request_ignore_err, BASE_URL + case_id) for case_id in ids]
    responses = [await f for f in futures]

    for response in responses:
        if response is None:
            continue
        parsed = parse_response(response)
        if parsed is not None:
            status, form_type = parsed
            counter[f'{center_name}|{two_digit_yr}|{day}|{code}|{form_type}|{status}'] += 1
    merge(counter)


def run():
    counter = Counter()
    for day in range(145, 200):
        for code in [5]:
            for center in CENTER_NAMES:
                loop = asyncio.get_event_loop()
                loop.run_until_complete(claw(center, 20, day, code, counter))


run()

File: src/clawer.ts
import * as cheerio from "cheerio";
import * as fs from "fs";
import * as https from "https";
import * as Immutable from "immutable";
import * as stringify from "json-stable-stringify";
import * as JSON5 from "json5";
import * as lodash from "lodash";
import type { Response, RequestInfo, RequestInit } from "node-fetch";
import fetch, { FetchError } from "node-fetch";
import nullthrows from "nullthrows";
import Constants from "./Constants";

https.globalAgent.options.rejectUnauthorized = false;

// const DATA_FILE_PATH = __dirname + "/data.json5";
const BASE_URL =
  "https://egov.uscis.gov/casestatus/mycasestatus.do?appReceiptNum=";
const today = Math.floor((new Date().getTime() - 3600 * 1000 * 7) / 86400000);

// 4 digit serial number
type CaseNumberFormat = "center-year-day-code-serial"  // i-797
  | "center-year-code-day-serial"; // i-485

const DATA_FILE_PATH: Map<CaseNumberFormat, string> = new Map([
  ["center-year-day-code-serial", __dirname + "/data.json5"],
  ["center-year-code-day-serial", __dirname + "/data485.json5"],
]);

class FetchPool {
  concurrency: number;
  running: number;
  constructor(concurrency: number) {
    this.concurrency = concurrency;
    this.running = 0;
  }

  async fetch(url: RequestInfo, init?: RequestInit): Promise<Response> {
    if (this.running > this.concurrency) {
      await this.sleep(Math.random() * 3000);
      return this.fetch(url, init);
    } else {
      this.running++;
      const result = fetch(url, init);
      this.running--;
      return result;
    }
  }

  sleep(waitTimeInMs: number) { new Promise(resolve => setTimeout(resolve, waitTimeInMs)); };
}
const fetchPool = new FetchPool(1000);

const getCaseID = (
  center_name: string,
  two_digit_yr: number,
  day: number,
  code: number,
  case_serial_numbers: number,
  case_number_format: CaseNumberFormat
): string => {
  switch (case_number_format) {
    case "center-year-code-day-serial":
      return center_name +
        two_digit_yr.toString() +
        code.toString() +
        day.toString().padStart(3, "0") +
        case_serial_numbers.toString().padStart(4, "0");
    case "center-year-day-code-serial":
      return center_name +
        two_digit_yr.toString() +
        day.toString().padStart(3, "0") +
        code.toString() +
        case_serial_numbers.toString().padStart(4, "0");
  }
};

const getStatus = async (
  url: string,
  retry: number = 6
): Promise<{ status: string; formType: string; } | null> => {
  if (retry <= 0) {
    console.log(`Request for ${url} failed too many times`);
    return null;
  }
  try {
    // 30 seconds timeout. always retry if timetout
    const f = await fetch(url, { timeout: 1000 * 30 });
    const t = await f.text();
    const status_regexp = new RegExp("(?<=<h1>).*(?=</h1>)");
    const status = nullthrows(
      status_regexp.exec(cheerio.load(t)(".text-center").html() as string)
    )[0];
    return status === ""
      ? null
      : {
        status,
        formType:
          Constants.FORM_TYPES.find((form) => t.includes(form)) ??
          "unknown form type",
      };
  } catch (e) {
    if (e instanceof FetchError && (e.message.includes('timeout') || e.message.includes('EADDRNOTAVAIL'))) {
      console.log('timeout! ' + url);
      return getStatus(url, retry - 1);
    } else {
      console.log(e + " " + url);
      return null;
    }
  }
};

const getLastCaseNumber = async (
  center_name: string,
  two_digit_yr: number,
  day: number,
  code: number,
  case_number_format: CaseNumberFormat
): Promise<number> => {
  let [low, high] = [1, 1];
  while (await getStatus(BASE_URL + getCaseID(center_name, two_digit_yr, day, code, high, case_number_format))) {
    [low, high] = [high, high * 2];
  }

  while (low < high) {
    const mid = Math.floor((low + high) / 2);
    const result = await getStatus(
      BASE_URL + getCaseID(center_name, two_digit_yr, day, code, mid, case_number_format)
    );
    if (result) {
      low = mid + 1;
    } else {
      high = mid;
    }
  }
  return low - 1;
};

const claw = async (
  center_name: string,
  two_digit_yr: number,
  day: number,
  code: number,
  format: CaseNumberFormat
): Promise<void> => {
  const path = DATA_FILE_PATH.get(format)!;
  const last = await getLastCaseNumber(center_name, two_digit_yr, day, code, format);
  if (last <= 0) {
    console.log(`No entires for ${center_name} day ${day}`);
    return;
  }

  console.log(`Loading ${last} entires for ${center_name} day ${day}`);
  // const results = (await PromisePool
  //   .withConcurrency(1000)
  //   .for(lodash.range(1, last + 1))
  //   .process(async case_number => await getStatus(BASE_URL + getCaseID(center_name, two_digit_yr, day, code, case_number, format))))
  //   .results
  //   .filter(Boolean)
  //   .map((x) => nullthrows(x));

  const results = (
    await Promise.all(
      lodash
        .range(1, last + 1)
        .map((case_number) =>
          getStatus(
            BASE_URL +
            getCaseID(center_name, two_digit_yr, day, code, case_number, format)
          )
        )
    )
  )
    .filter(Boolean)
    .map((x) => nullthrows(x));

  const counter = results
    .reduce((counter, res) => {
      const key = `${center_name}|${two_digit_yr}|${day}|${code}|${res.formType}|${res.status}`;
      return counter.set(key, 1 + (counter.get(key) ?? 0));
    }, Immutable.Map<string, number>())
    .map((val) => ({ [today]: val }))
    .toObject();

  const json5_obj = JSON5.parse(
    fs.readFileSync(path, { encoding: "utf8" })
  );

  const new_json5_obj = { ...json5_obj };

  Object.entries(counter).forEach(([key, count]) => {
    new_json5_obj[key] = { ...(new_json5_obj[key] ?? {}), ...count };
  });

  Object.entries(new_json5_obj).forEach(([key, count]) => {
    new_json5_obj[key] = lodash.pickBy(new_json5_obj[key], (_count, day) => Math.abs(Number.parseInt(day) - today) <= 14);
    if (lodash.isEmpty(new_json5_obj[key])) {
      delete new_json5_obj[key];
    }
  });

  fs.writeFileSync(
    path,
    // @ts-ignore: solve export issue for json stable stringify
    JSON5.stringify(JSON5.parse(stringify(new_json5_obj)), {
      space: 2,
      quote: '"',
    }),
    { encoding: "utf8" }
  );
  console.log(`Finished ${last} entires for ${center_name} day ${day}`);
};

(async () => {
  for (const d of lodash.range(1, 350)) {
    if (process.argv.includes("485")) {
      for (const name of Constants.CENTER_NAMES) {
        await claw(name, 21, d, 9, 'center-year-code-day-serial');
      }
    } else {
      await Promise.all(
        Constants.CENTER_NAMES.map((name) => claw(name, 21, d, 5, 'center-year-day-code-serial'))
      );
    }
  }
})();
